### ISSUE TITLE
Add a link between images and resources being deployed

### ISSUE THREAD

**Issue by Aaron-9900** [argoproj/argo-cd#20896]
# Summary

We (Datadog) have a product that uses Argo CD notifications in order to know when an Argo CD sync has happened, and provide additional observability to customers on those deployments inside Datadog. We would like to receive the images being deployed by each of the resources in an ArgoCD application on its notification webhook. Currently the notification sends an "images" array, but its not possible to link the images with the resources it relates to.

# Motivation

In Datadog we are working on a [CD Visibility product](https://docs.datadoghq.com/continuous_delivery/deployments/argocd/) that provides insights into the deployments of customers, ArgoCD being one of the providers we support. One of the key pieces of information we would like to provide is the changes that occurred between one deployment and another. We are currently able to get the changed resources by looking at the sync message of each resource, yet currently there is no visibility into how the image changed for those resources. We think that providing this information is valuable not only for us in CD Visibility, but to any user consuming ArgoCD Notifications.

# Proposal

1. In [`SyncAppState`](https://github.com/argoproj/argo-cd/blob/master/controller/sync.go#L95) add the images to the `ResourceResult` struct. I think we would need access to `ApplicationTree` in `SyncAppState`, although maybe there is a better way.
2. Modify the CRD so that `ResourceResult` includes an Images array

Another possibility would be to modify `GetSummary` to add the images to the resources.

If the use case and the proposed solution seem reasonable, we are happy to contribute ourselves to the repository!

**Comment by andrii-korotkov-verkada** [comment]
Seems reasonable to me.

**Comment by andrii-korotkov-verkada** [comment]
@Aaron-9900, should I assign this to you?

**Comment by Aaron-9900** [comment]
@andrii-korotkov-verkada sure!

**Comment by Aaron-9900** [comment]
@andrii-korotkov-verkada  Here is the PR in gitops engine: https://github.com/argoproj/gitops-engine/pull/642

Thank you 🙂

**Comment by Aaron-9900** [comment]
Hey @crenshaw-dev, I think I can answer all of those points:

> CRD size: the image list size is unbounded, the sync operation field could start to get problematically big

The image list is already present in the CRD, so even if it’s unbounded, this issue already exists in the current state of ArgoCD. This is about adding a link between those images and the resources that hold them.


> CPU: there aren't any benchmarks on the gitops-engine PR, so it's unclear whether retrieving images will be costly

I’ll run benchmarks if it helps build confidence in getting this merged.


> Imprecision: the code currently handles only a few known k8s native types; what about other workloads like Argo Workflows?

This isn’t really an issue of imprecision but rather completeness. We can’t handle all possible CRDs, but the current implementation already covers native types along with multiple CRDs that follow their pattern, such as Argo Rollouts. I can certainly make the necessary changes to support Argo Workflows as well.


> Argo CD use case: how will Argo CD users benefit from this feature outside its Datadog application? The images will be available on the App resource, but they currently won't be surfaced in any useful way as far as I can tell.

While this originated from Datadog, I believe it’s a significant step toward a more powerful notification system in ArgoCD. Right now, you can determine whether a deployment succeeded or not, but there’s no easy way to know:
1. What resources changed 
2. How they changed
This change addresses point 2 by providing image-related information, which will be available through webhooks or any other consumer in the notifications engine—not just Datadog. While we use webhooks, any consumer can leverage this functionality.



Point 1 remains a challenge. Currently, consumers can rely on the “message” field of the resource object to detect changes by checking for terms like “configured” or “unchanged.” However, this approach is not foolproof—especially when Server-Side Apply is used. In the future, we plan to improve visibility into changed resources from the Argo CD side, ensuring more reliable change detection.


Addressing both points will significantly enhance notifications, making them more precise and actionable. For example, an Argo CD application managing multiple services within a monorepo could enable targeted notifications, ensuring that users are alerted only when specific resources change, and how they changed.

**Comment by crenshaw-dev** [comment]
> The image list is already present in the CRD

Ah yep. That list is deduped across the whole app, and the new field would be per-resource. So there's some additional storage but probably not wildly different.

> I’ll run benchmarks if it helps build confidence in getting this merged.

It would!

> This isn’t really an issue of imprecision but rather completeness.

There's also the possibility that some resource matches the expected pattern but doesn't actually run the specified image. The ApplicationSummary images list is based only on Pods, which is a much stronger source of truth. (Of course, you don't have access to these Pods at sync time, so I understand why we're not using the same code.)

> it’s a significant step toward a more powerful notification system in ArgoCD

That makes sense. Ideally I'd like to make sure this is a feature Argo CD OSS users will find useful (there's a chance there are already feature requests for it on this repo). I'd also want docs and/or tests showing how to take advantage of the new feature without involving 3rd party tools.

tl;dr I'm not strongly opposed to the change, I just want to make sure we're providing enough value to the OSS users to justify the costs.

**Comment by andrii-korotkov-verkada** [comment]
I think that's something my company may find useful. We haven't scoped the tooling yet, but this might be relevant especially for regulated environments like Fed.

**Comment by rumstead** [comment]
We currently have a 1-1 mapping, generally speaking, of an image to an application and are getting away with the existing image list. We use Argo webhooks to signal to our change management systems that an image release happened. We would take advantage of the mapping for our applications that are "multi-image".

**Comment by crenshaw-dev** [comment]
If y'all have particular notifications configs you'd use with this feature, would you mind posting them? That'll give us something to test/validate against.

**Comment by Aaron-9900** [comment]
Here is the notifications-cm we ask our customers to have in DataDog:
<details>

```yaml
apiVersion: v1
kind: ConfigMap
metadata:
  name: argocd-notifications-cm
data:
  service.webhook.cd-visibility-webhook: |
    url: https://webhook-intake.datadoghq.com/api/v2/webhook
    headers:
    - name: "DD-CD-PROVIDER-ARGOCD"
      value: "true"
    - name: "DD-API-KEY"
      value: $dd-api-key
    - name: "Content-Type"
      value: "application/json"    
  template.cd-visibility-template: |
    webhook:
      cd-visibility-webhook:
        method: POST
        body: |
            {
              "app": {{toJson .app}},
              "context": {{toJson .context}},
              "service_type": {{toJson .serviceType}},
              "recipient": {{toJson .recipient}},
              "commit_metadata": {{toJson (call .repo.GetCommitMetadata .app.status.operationState.syncResult.revision)}}
            }    
  trigger.cd-visibility-trigger: |
    - when: app.status.operationState.phase in ['Succeeded', 'Failed', 'Error'] and app.status.health.status in ['Healthy', 'Degraded']
      send: [cd-visibility-template]
    - when: app.status.operationState.phase == 'Running' and app.status.health.status in ['Healthy', 'Degraded']
      send: [cd-visibility-template]    
```
<details>

But what I think is more relevant is what happens when receiving the notification: we parse the webhook to find changed resources and (ideally with this issue) the images related.

**Comment by Aaron-9900** [comment]
> Id also want docs and/or tests showing how to take advantage of the new feature without involving 3rd party tools.

I've been looking at this and I don't see a place in the docs where this currently fits well. I've though of adding a new section called "Guides"/"Examples" (or something similar) inside "Notifications" where we can add different use-cases for notifications in ArgoCD, such as this one. WDYT?

**Comment by Aaron-9900** [comment]
👋🏼 Got the Docs & GitOps PRs approved. What are the next steps to get this merged? Thanks!

**Comment by crenshaw-dev** [comment]
Apologies for the delay!

One last request for the PR if possible: can we add the list of images as a component in the UI for resource details? I'd like to leverage the feature outside just notifications, if only to increase awareness that the feature exists.

**Comment by Aaron-9900** [comment]
> One last request for the PR if possible: can we add the list of images as a component in the UI for resource details? I'd like to leverage the feature outside just notifications, if only to increase awareness that the feature exists.

Makes sense, I'll prepare a PR for it

**Comment by Aaron-9900** [comment]
@crenshaw-dev Done in [feat: show operation state UI](https://github.com/argoproj/argo-cd/pull/23589)
